### ISSUE TITLE
Beam search issue

### ISSUE THREAD

**Issue by HichemAK** [uiuc-focal-lab/syncode#104]
Hello,

It seems that beam search is not supported at the moment

### Code to reproduce error

```python
from syncode import SyncodeLogitsProcessor, Grammar
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

grammar = """ start: " " city "."
city: "Paris" | "Berlin" | "Madrid" | "Marseille" | "Algeirs" | "Dublin" | "Seattle"         
"""

model_name = "EleutherAI/pythia-14m"

# Load the unconstrained original model
model = AutoModelForCausalLM.from_pretrained(model_name, torch_dtype=torch.bfloat16, device_map='cuda')
tok = AutoTokenizer.from_pretrained(model_name)

grammar = Grammar(grammar)
syncode_logits_processor = SyncodeLogitsProcessor(grammar=grammar, tokenizer=tok, parse_output_only=True)

prompt = 'The capital of France is'
syncode_logits_processor.reset(prompt)
inputs = tok([prompt], return_tensors='pt')
output = model.generate(
          inputs.input_ids.cuda(),
          max_length=50, 
          num_return_sequences=2,
          num_beams=2,
          pad_token_id=tok.eos_token_id, 
          logits_processor=[syncode_logits_processor]
        )

[print(tok.decode(x)) for x in output]
# Output
# The capital of France is A, "1, '1, '1, "1, '0',  '0!
# <|endoftext|><|endoftext|><|endoftext|><|endoftext|>
# The capital of France is A, "1, '1, '1, "1, '0',  '0!
# "1"<|endoftext|>
```

We can see from the output that the grammar isn't followed by the model while it works perfectly when num_beams=1 and num_return_sequences=1

**Comment by shubhamugare** [comment]
Hi @HichemAK,

Looking into this, will get back to you soon.

**Comment by shubhamugare** [comment]
Hi @HichemAK,

At this point, the `SyncodeLogitsProcessor` needs an argument `num_samples=2` to work correctly. (Which I'm not sure if is 100% necessary)

 I added an assertion with the appropriate error message to crash the program if it is not provided in PR #105. Please check this [notebook](https://github.com/uiuc-focal-lab/syncode/blob/beam_search/notebooks/tests/beam_search.ipynb) to see how the output looks like for your example. Also, I switched SynCode to the `grammar_strict` mode to get a perfect output.

**Comment by teoremma** [comment]
Hello @shubhamugare 

It seems this bug is still not fully resolved, particularly if `do_sample=True`, even for `num_samples=2`.

### Code to reproduce the error
```python
from syncode import SyncodeLogitsProcessor, Grammar
from transformers import AutoModelForCausalLM, AutoTokenizer
import lark

grammar_str = r"""
start: item ("," item)* 

item: "'" name "'"
    | "\"" name "\""

name: "Alice" 
    | "Bob" 
    | "Carol" 
    | "Dave"
    | "Eve"
"""

device = "cuda"
model_name = "deepseek-ai/deepseek-coder-1.3b-base"
model = AutoModelForCausalLM.from_pretrained(model_name, device_map="auto").eval()
tokenizer = AutoTokenizer.from_pretrained(model_name)

syncode_grammar = Grammar(grammar_str)
parser = lark.Lark(grammar_str)

prompt = "A list of male first names:\n"

inputs = tokenizer(prompt, return_tensors="pt").input_ids.to(device)

constrain = True

args = {
    "max_new_tokens" : 128,
    "do_sample" : True,
    "num_beams" : 2,
    "num_return_sequences" : 2,
    "pad_token_id" : tokenizer.eos_token_id,
}

syncode_logits_processor = SyncodeLogitsProcessor(
    grammar=syncode_grammar, 
    tokenizer=tokenizer, 
    parse_output_only=True, 
    num_samples=args["num_beams"],
    mode="grammar_strict",
)
syncode_logits_processor.reset(prompt)

outputs = model.generate(
    inputs,
    logits_processor=[syncode_logits_processor] if constrain else None,
    **args,
)

outputs = [o[len(inputs[0]):] for o in outputs]
completions = tokenizer.batch_decode(outputs, skip_special_tokens=True)
completions_tokens = [[tokenizer.decode(tok) for tok in output] for output in outputs]
print(completions)


for i, (c, toks) in enumerate(zip(completions, completions_tokens)):
    print(f">>> COMPLETION {i}\n")
    try:
        tree = parser.parse(c)
        print("CAN PARSE\n")
    except:
        print("CANNOT PARSE\n") 
    print(f"{c}\n")
    print(f"{toks}\n")
```

Around 1/3 of the time, the model will generate invalid strings.
When that happens, the resulting string has some non-empty prefix of tokens that would be valid, for example

```
# string
'"Aaron", "Allen", "Albert", ...

# tokens
['"', 'A', 'aron', '",', ...
```

It leads me to think that the bug must be somewhere in the lexer processing.

We are currently blocked since we depend on beam search, both with greedy and sampling decoding, for `n > 2` so please look into this issue ASAP.

Thank you.

**Comment by shubhamugare** [comment]
Hi @teoremma ,

Thanks for pointing out the issue. 

I see what's happening here, I found the problem is how we handle incrementality in the parser. For each sequence that is generated, we maintain one incremental parser that restores the parser state. During beam search, these sequences are reordered and that can cause random behavior. Removing this part [here](https://github.com/uiuc-focal-lab/syncode/blob/ba2e5e94e53e8df0559c120977ea725f63edd029/syncode/parsers/incremental_parser.py#L126) temporarily fixes the problem at the cost of efficiency. It is hard to say exactly how slow it will be, but it will be slower if you are generating longer sequences. 

For a slightly longer-term fix, I think the incremental parser needs to be indexed differently than how it works now and we shouldn't have a separate instance of parser for each generation. Ideally, we should have one global incremental parser that stores all previous parser states.

**Comment by shubhamugare** [comment]
Hi @teoremma,

This should be resolved in PR #113. I'm including your notebook in the test notebooks. Can you confirm, if it looks good?

**Comment by teoremma** [comment]
It seems to be working on my end, thanks a lot!
